core/core-impl/src/main/java/ai/timefold/solver/core/impl/score/director/AbstractScoreDirector.java

```diff
@@ -644,53 +693,166 @@ public void assertExpectedUndoMoveScore(Move<Solution_> move, Score_ beforeMoveS
+                    "after", expectedBeforeAfterCalls, afterVariableChangedUndoEvents,
+                    VIOLATION_LIMIT);
+        }
+
+        if (out.isEmpty()) {
+            return "Genuine and shadow variables agree with from scratch calculation after the undo move and match the state prior to the move.";
+        }
+
+        return out.toString();
+    }
+
+    private void appendVariableChangedViolations(StringBuilder out, String prefix,
+            List<Pair<VariableDescriptor<Solution_>, Object>> changedVariables,
+            AllVariablesAssert<Solution_> expectedSnapshot,
+            AllVariablesAssert<Solution_> actualSnapshot, int limit) {
+        out.append(prefix);
+        out.append(":\n");
+        int violationCount = 0;
+        for (var changedVariable : changedVariables) {
+            var expectedSnapshotVariable = expectedSnapshot.getVariableSnapshot(changedVariable);
+            var actualSnapshotVariable = actualSnapshot.getVariableSnapshot(changedVariable);
+            out.append("    ");
+            out.append(expectedSnapshotVariable.getVariableDescriptor().getSimpleEntityAndVariableName());
+            out.append(" (");
+            out.append(expectedSnapshotVariable.getEntity());
+            out.append(") expected (");
+            out.append(expectedSnapshotVariable.getValue());
+            out.append(") actual (");
+            out.append(actualSnapshotVariable.getValue());
+            out.append(")\n");
+            violationCount++;
+            if (violationCount >= limit) {
+                return;
+            }
+        }
+    }
 
-        String undoShadowVariableViolations = afterMoveSnapshot.createShadowVariablesViolationMessage(3L);
-        String beforeShadowVariableViolations = null;
-
-        if (beforeMoveSnapshot != null) {
-            beforeShadowVariableViolations = beforeMoveSnapshot.createShadowVariablesViolationMessage(3L);
-        }
-        if (undoShadowVariableViolations != null && beforeShadowVariableViolations != null) {
-            return "Shadow variables have different values when recalculated from scratch before and after undo:\n" +
-                    "Before undo: " + beforeShadowVariableViolations + "\n" +
-                    "After undo: " + undoShadowVariableViolations + "\n";
-        } else if (undoShadowVariableViolations != null) {
-            return "Shadow variables have different values when recalculated from scratch after undo:\n" +
-                    undoShadowVariableViolations + "\n";
-        } else if (beforeShadowVariableViolations != null) {
-            return "Shadow variables have different values when recalculated from scratch before undo:\n"
-                    + beforeShadowVariableViolations
-                    + "\n";
-        } else {
-            return "Shadow variables agrees with from scratch calculations before and after undo.\n";
+    private void appendVariableChangedEventViolations(StringBuilder out, String prefix,
```

Comment:
Let's make this a function without side effects; return `String`, don't accept `out`.